availability-store/src/worker.rs

```diff
@@ -0,0 +1,989 @@
+		for (candidate, block) in blocks.into_iter() {
+			let _ = self.availability_store.add_candidate(&candidate);
+
+			if let Some((_block, _msgs)) = block {
+				// Should we be breaking block into chunks here and gossiping it and so on?
+			}
+
+			if let Err(e) = self.register_chunks_listener(
+				runtime_handle,
+				sender,
+				relay_parent,
+				candidate.erasure_root
+			) {
+				warn!(target: LOG_TARGET, "Failed to register chunk listener: {}", e);
+			}
+		}
+
+		let _ = self.availability_store.add_candidates_in_relay_block(
+			&relay_parent,
+			hashes
+		);
+
+		Ok(())
+	}
+
+	// Processes chunks messages that contain awaited items.
+	//
+	// When an awaited item is received, it is placed into the availability store
+	// and removed from the frontier. Listener de-registered.
+	fn on_chunks_received(
+		&mut self,
+		relay_parent: Hash,
+		candidate_hash: Hash,
+		chunks: Vec<ErasureChunk>
+	) -> Result<(), Error> {
+		let (_, n_validators) = self.availability_store
+			.get_validator_index_and_n_validators(&relay_parent)
+			.ok_or(Error::IdAndNValidatorsNotFound { relay_parent })?;
+
+		let receipt = self.availability_store.get_candidate(&candidate_hash)
+			.ok_or(Error::CandidateNotFound { candidate_hash })?;
+
+		for chunk in &chunks {
+			let topic = erasure_coding_topic(relay_parent, receipt.erasure_root, chunk.index);
+			// need to remove gossip listener and stop it.
+			if let Some(signal) = self.registered_gossip_streams.remove(&topic) {
+				signal.fire();
+			}
+		}
+
+		self.availability_store.add_erasure_chunks(
+			n_validators,
+			&relay_parent,
+			&candidate_hash,
+			chunks,
+		)?;
+
+		Ok(())
+	}
+
+	// Adds the erasure roots into the store.
+	fn on_erasure_roots_received(
+		&mut self,
+		relay_parent: Hash,
+		erasure_roots: Vec<Hash>
+	) -> Result<(), Error> {
+		self.availability_store.add_erasure_roots_in_relay_block(&relay_parent, erasure_roots)?;
+
+		Ok(())
+	}
+
+	// Processes the `ListenForChunks` message.
+	//
+	// When the worker receives a `ListenForChunk` message, it double-checks that
+	// we don't have that piece, and then it registers a listener.
+	fn on_listen_for_chunks_received(
+		&mut self,
+		runtime_handle: &mut Handle,
+		sender: &mut mpsc::UnboundedSender<WorkerMsg>,
+		relay_parent: Hash,
+		candidate_hash: Hash,
+		id: usize
+	) -> Result<(), Error> {
+		let candidate = self.availability_store.get_candidate(&candidate_hash)
+			.ok_or(Error::CandidateNotFound { candidate_hash })?;
+
+		if self.availability_store
+			.get_erasure_chunk(&relay_parent, candidate.block_data_hash, id)
+			.is_none() {
+			if let Err(e) = self.register_chunks_listener(
+				runtime_handle,
+				sender,
+				relay_parent,
+				candidate.erasure_root
+			) {
+				warn!(target: LOG_TARGET, "Failed to register a gossip listener: {}", e);
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn start(mut self) -> WorkerHandle {
+		let sender = self.sender.clone();
+		let mut receiver = self.receiver.take()
```

Comment:
The issue here seems to be with calling `start` twice. Why is the worker instantiated with `Worker::new(...).start`, rather than a free ` fn start(...) -> Worker`?